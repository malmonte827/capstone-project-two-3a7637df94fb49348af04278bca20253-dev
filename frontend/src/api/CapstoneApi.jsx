import axios from "axios"

const BASE_URL = process.env.React_APP_BASE_URL || "http://localhost:3001"

/** APi class
 * 
 * Staic class to get/send to the API
 */

class CapstoneApi{
    // token for interacting with the API
    static token

    static async request(endpoint, data={}, method="get"){
        console.debug(`API Call:`, endpoint, data, method)

        const url = `${BASE_URL}/${endpoint}`
        const headers = {Authorization: `Bearer ${CapstoneApi.token}`}
        const params = (method === "get") ? data : {}

        try{
            return (await axios({url, method, data, params,headers})).data
        }catch(err){
            console.error(`API Error:`, err.response)
            let message = err.response.data.error.message
            throw Array.isArray(message) ? message : [message]
        }
    }

    /** API ROUTES */

    /** Get the current user */

    static async getCurrentUser(username){
        let res = await this.request(`/users/${username}`)
        return res.user
    }

/** Signup User */
    static async signup(data){
        let res = await this.request(`/auth/register`, data, "post")
        return res.token
    }

    /** Login User */
    static async login(data){
        const res = await this.request(`/auth/token`, data, "post")
        return res.token
    }
}

export default CapstoneApi