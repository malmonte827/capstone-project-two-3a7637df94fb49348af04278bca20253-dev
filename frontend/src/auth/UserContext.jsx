import React from "react"

/** COntext provides current user object throughout app */

const UserContext = React.createContext()

export default UserContext