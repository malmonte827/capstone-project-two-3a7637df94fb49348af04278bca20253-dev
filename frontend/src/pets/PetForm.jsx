import React, { useState, useContext } from "react";
import { useNavigate, useLocation, useParams } from "react-router-dom";
import Alert from "../common/Alert"
import "./PetForm.css"
import CapstoneApi from "../api/CapstoneApi";
import UserContext from "../auth/UserContext";

/**  PetForm Component
 *
 * Handles adding a new pet or updating an exsisting pets info 
 * 
 */

function PetForm() {
    const location = useLocation()
    const {pet} = location.state || {}
    const {currentUser} = useContext(UserContext)

    const [formData, setFormData] = useState({
        name: pet ? pet.name :  "",
        age: pet ? pet.age :  "",
        species: pet ? pet.species :  "",
        hunger: pet ? pet.hunger :  ""
    });

    const [formErrors, setFormErrors] = useState([]);

    const navigate = useNavigate()

    /** 
     * Handles form input changes.
     * Updates the corresponding form field in state.
     *
     * @param {Object} evt - Event object from input field change.
     */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData((data) => ({ ...data, [name]: value }));
    }

    /**
     * Handles form submission.
     *
     * If a pet exists updates the pets info via an udatePet call
     * If no pet exists it creates a new pet via createPet call
     * Redirects to /pets upon success.
     *
     * @param {Object} evt - Form submission event.
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        try {
            if(pet){
                console.log("in update")
                await CapstoneApi.updatePet(currentUser.username, pet.pet_id, formData)
                navigate("/pets")
            }else{
                
                console.log("in create")
                const updatedFormData = {...formData, user_id:currentUser.id}
                console.log(updatedFormData)
                await CapstoneApi.createPet(currentUser.username, updatedFormData);
                navigate("/pets");
            }
        } catch (err) {
            setFormErrors(err);
        }
    }

    return (
        <div className="PetForm">
            <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
                <h2 className="mb-3">{ pet ? "Edit Pet" : "Add Pet"}</h2>
                <div className="card">
                    <div className="card-body">
                        <form onSubmit={handleSubmit}>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="name"
                                >
                                    Name
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    id="name"
                                    name="name"
                                    value={formData.name}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="species"
                                >
                                    Species
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    id="species"
                                    name="species"
                                    value={formData.species}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="age"
                                >
                                    Age
                                </label>
                                <input
                                    className="form-control"
                                    type="number"
                                    id="age"
                                    name="age"
                                    value={formData.age}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="hunger"
                                >
                                    Hunger
                                </label>
                                <input
                                    className="form-control"
                                    type="int"
                                    id="hunger"
                                    name="hunger"
                                    value={formData.hunger}
                                    onChange={handleChange}
                                />
                            </div>

                            {formErrors.length ? (
                                <Alert type="danger" messages={formErrors} />
                            ) : null}
                            <div className="d-grid">
                                <button className="btn btn-primary">
                                    Submit
                                </button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    );
}

export default PetForm;
