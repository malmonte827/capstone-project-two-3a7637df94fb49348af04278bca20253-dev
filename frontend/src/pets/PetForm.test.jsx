import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { expect, vi } from "vitest";
import PetForm from "./PetForm";
import UserContext from "../auth/UserContext";
import CapstoneApi from "../api/CapstoneApi";

vi.mock("../api/CapstoneApi");

describe("<PetForm />", () => {
    const currentUser = { username: "testuser", id: 1 };

    it("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetForm />
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetForm />
                </UserContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("handles form input changes", () => {
        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetForm />
                </UserContext.Provider>
            </MemoryRouter>
        );

        const nameInput = screen.getByLabelText("Name");
        fireEvent.change(nameInput, { target: { value: "Fluffy" } });
        expect(nameInput.value).toBe("Fluffy");
    });

    it("submits the form and calls createPet API", async () => {
        CapstoneApi.createPet.mockResolvedValueOnce({});

        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetForm />
                </UserContext.Provider>
            </MemoryRouter>
        );

        fireEvent.change(screen.getByLabelText("Name"), {
            target: { value: "Fluffy" },
        });
        fireEvent.change(screen.getByLabelText("Species"), {
            target: { value: "Cat" },
        });
        fireEvent.change(screen.getByLabelText("Age"), {
            target: { value: "2" },
        });
        fireEvent.change(screen.getByLabelText("Hunger"), {
            target: { value: "5" },
        });
        fireEvent.change(screen.getByLabelText("Description"), {
            target: { value: "A cute cat" },
        });

        fireEvent.click(screen.getByText("Submit"));

        await waitFor(() => {
            expect(CapstoneApi.createPet).toHaveBeenCalledWith("testuser", {
                name: "Fluffy",
                species: "Cat",
                age: "2",
                hunger: 5,
                description: "A cute cat",
                user_id: 1,
            });
        });
    });

    it("handles API errors and displays an alert", async () => {
        CapstoneApi.createPet.mockRejectedValueOnce(["Error creating pet"]);

        render(
            <MemoryRouter>
                <UserContext.Provider value={{ currentUser }}>
                    <PetForm />
                </UserContext.Provider>
            </MemoryRouter>
        );

        fireEvent.click(screen.getByText("Submit"));

        await waitFor(() => {
            expect(screen.getByText("Error creating pet")).toBeInTheDocument();
        });
    });
});
