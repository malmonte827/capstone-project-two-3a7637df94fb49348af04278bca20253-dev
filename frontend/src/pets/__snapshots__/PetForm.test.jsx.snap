// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PetForm /> > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PetForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h2
        class="mb-3"
      >
        Add Pet
      </h2>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="name"
              >
                Name
              </label>
              <input
                class="form-control"
                id="name"
                name="name"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="species"
              >
                Species
              </label>
              <input
                class="form-control"
                id="species"
                name="species"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="age"
              >
                Age
              </label>
              <input
                class="form-control"
                id="age"
                name="age"
                type="number"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="hunger"
              >
                Hunger
              </label>
              <input
                class="form-control"
                id="hunger"
                name="hunger"
                type="number"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="description"
              >
                Description
              </label>
              <textarea
                class="form-control"
                id="description"
                maxlength="250"
                name="description"
              />
            </div>
            <div
              class="d-grid"
            >
              <button
                class="btn btn-primary"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
